plugins/woocommerce/tests/php/includes/wc-order-functions-test.php

```diff
@@ -56,4 +56,59 @@ public function test_wc_restock_refunded_items_stock_metadata() {
 		$this->assertEquals( 0, (int) $line_item->get_meta( '_reduced_stock', true ) );
 		$this->assertEquals( 2, (int) $line_item->get_meta( '_restock_refunded_items', true ) );
 	}
+
+	/**
+	 * Test update_total_sales_counts and check total_sales after order reflection.
+	 *
+	 * Tests the fix for issue #23796
+	 */
+	public function test_wc_update_total_sales_counts() {
+
+		$product = WC_Helper_Product::create_simple_product();
+
+		WC()->cart->add_to_cart( $product->get_id() );
+
+		$order_id = WC_Checkout::instance()->create_order(
+			array(
+				'billing_email'  => 'a@b.com',
+				'payment_method' => 'dummy',
+			)
+		);
+
+		$this->assertEquals( 0, $product->get_total_sales() );
+
+		$order = new WC_Order( $order_id );
+
+		$order->update_status( 'processing' );
+		$this->assertEquals( 1, $product->get_total_sales() );
+
+		$order->update_status( 'cancelled' );
+		$this->assertEquals( 0, $product->get_total_sales() );
+
+		$order->update_status( 'processing' );
+		$this->assertEquals( 1, $product->get_total_sales() );
+
+		$order->update_status( 'completed' );
+		$this->assertEquals( 1, $product->get_total_sales() );
+
+		$order->update_status( 'refunded' );
+		$this->assertEquals( 1, $product->get_total_sales() );
+
+		$order->update_status( 'processing' );
+		$this->assertEquals( 1, $product->get_total_sales() );
+
+		// Test order trash / un-trash.
+		if ( $order->delete( false ) ) {
+			$this->assertEquals( 0, $product->get_total_sales() );
+
+			$order->untrash();
+			$this->assertEquals( 1, $product->get_total_sales() );
+		}
+
+		// Test force delete.
+		if ( $order->delete( true ) ) {
+			$this->assertEquals( 0, $product->get_total_sales() );
+		}
+	}
```

Comment:
I'd propose a few changes to get this test passing:

- Once we have an instance of a product object, its knowledge (at least in relation to the total sales count) is 'frozen', so we need to grab fresh instances to check the new value.
- An interesting aspect of order objects is that, following a delete (whether soft or hard), the order ID is set to zero, and so again there are points where we need to obtain and work with fresh instances of the order object.
- I understand why we're nesting some of the assertions in if statements, but flattening this out A) arguably improves readability B) prevents silent failures (such as if execution does not enter the if block).

Here's an iteration which I think clears the current failing checks:

```suggestion
	public function test_wc_update_total_sales_counts() {

		$product_id = WC_Helper_Product::create_simple_product()->get_id();

		WC()->cart->add_to_cart( $product_id );

		$order_id = WC_Checkout::instance()->create_order(
			array(
				'billing_email'  => 'a@b.com',
				'payment_method' => 'dummy',
			)
		);

		$this->assertEquals( 0, wc_get_product( $product_id )->get_total_sales() );

		$order = new WC_Order( $order_id );

		$order->update_status( 'processing' );
		$this->assertEquals( 1, wc_get_product( $product_id )->get_total_sales() );

		$order->update_status( 'cancelled' );
		$this->assertEquals( 0, wc_get_product( $product_id )->get_total_sales() );

		$order->update_status( 'processing' );
		$this->assertEquals( 1, wc_get_product( $product_id )->get_total_sales() );

		$order->update_status( 'completed' );
		$this->assertEquals( 1, wc_get_product( $product_id )->get_total_sales() );

		$order->update_status( 'refunded' );
		$this->assertEquals( 1, wc_get_product( $product_id )->get_total_sales() );

		$order->update_status( 'processing' );
		$this->assertEquals( 1, wc_get_product( $product_id )->get_total_sales() );

		// Test trashing the order.
		$order->delete( false );
		$this->assertEquals( 0, wc_get_product( $product_id )->get_total_sales() );

		// To successfully untrash, we need to grab a new instance of the order.
		wc_get_order( $order_id )->untrash();
		$this->assertEquals( 1, wc_get_product( $product_id )->get_total_sales() );

		// Test full deletion of the order (again, we need to grab a new instance of the order).
		wc_get_order( $order_id )->delete( true );
		$this->assertEquals( 0, wc_get_product( $product_id )->get_total_sales() );
	}
```